src/main/java/com/spotify/confidence/eventsender/EventSenderEngineImpl.java

```diff
@@ -0,0 +1,123 @@
+package com.spotify.confidence.eventsender;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.concurrent.*;
+
+class EventSenderEngineImpl implements EventSenderEngine {
+  private final ExecutorService writeThread = Executors.newSingleThreadExecutor();
+  private final ExecutorService uploadThread = Executors.newSingleThreadExecutor();
+  private final BlockingQueue<Event> writeQueue = new LinkedBlockingQueue<>();
+  private final BlockingQueue<String> uploadQueue = new LinkedBlockingQueue<>();
+  private final BlockingQueue<String> shutdownQueue = new LinkedBlockingQueue<>(1);
+  private final EventSenderStorage eventStorage = new InMemoryStorage();
+  private final EventUploader eventUploader;
+  private final List<FlushPolicy> flushPolicies;
+  private static final String UPLOAD_SIG = "UPLOAD";
+  private static final String SHUTDOWN_UPLOAD = "SHUTDOWN_UPLOAD";
+  private static final String SHUTDOWN_WRITE = "SHUTDOWN_WRITE";
+  private boolean isStopped = false;
+
+  public EventSenderEngineImpl(List<FlushPolicy> flushPolicyList, EventUploader eventUploader) {
+    this.flushPolicies = flushPolicyList;
+    this.eventUploader = eventUploader;
+    writeThread.submit(new WritePoller());
+    uploadThread.submit(new UploadPoller());
+  }
+
+  class WritePoller implements Runnable {
+    @Override
+    public void run() {
+      while (true) {
+        try {
+          final Event event = writeQueue.take();
+          eventStorage.write(event);
+          flushPolicies.forEach(FlushPolicy::hit);
+
+          if (flushPolicies.stream().anyMatch(FlushPolicy::shouldFlush)) {
+            flushPolicies.forEach(FlushPolicy::reset);
```

Comment:
Maybe move this to last line in block so reset is not called until the batch is sent to the uploadQueue. I guess we don't expect to survive errors here, but still..